gcp/observability/observability_test.go

```diff
@@ -487,3 +526,541 @@ func (s) TestStartErrorsThenEnd(t *testing.T) {
+		t.Fatalf("error setting up observability %v", err)
+	}
+	defer cleanup()
+	ss := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, in *grpc_testing.SimpleRequest) (*grpc_testing.SimpleResponse, error) {
+			return &grpc_testing.SimpleResponse{}, nil
+		},
+		FullDuplexCallF: func(stream grpc_testing.TestService_FullDuplexCallServer) error {
+			_, err := stream.Recv()
+			if err != io.EOF {
+				return err
+			}
+			return nil
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Spawn a goroutine to receive the trace and span ids received by the
+	// exporter corresponding to a Unary RPC.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	readerErrCh := testutils.NewChannel()
+	unaryDone := grpcsync.NewEvent()
+	go func() {
+		defer wg.Done()
+		var traceAndSpanIDs []traceAndSpanID
+		val, err := idCh.Receive(ctx)
+		if err != nil {
+			readerErrCh.Send(fmt.Errorf("error while waiting for IDs: %v", err))
+		}
+
+		tasi, ok := val.(traceAndSpanID)
+		if !ok {
+			readerErrCh.Send(fmt.Errorf("received wrong type from channel: %T", val))
+		}
+		traceAndSpanIDs = append(traceAndSpanIDs, tasi)
+
+		val, err = idCh.Receive(ctx)
+		if err != nil {
+			readerErrCh.Send(fmt.Errorf("error while waiting for IDs: %v", err))
+		}
+
+		tasi, ok = val.(traceAndSpanID)
+		if !ok {
+			readerErrCh.Send(fmt.Errorf("received wrong type from channel: %T", val))
+		}
+		traceAndSpanIDs = append(traceAndSpanIDs, tasi)
+
+		val, err = idCh.Receive(ctx)
+		if err != nil {
+			readerErrCh.Send(fmt.Errorf("error while waiting for IDs: %v", err))
+		}
+		tasi, ok = val.(traceAndSpanID)
+		if !ok {
+			readerErrCh.Send(fmt.Errorf("received wrong type from channel: %T", val))
+		}
+		traceAndSpanIDs = append(traceAndSpanIDs, tasi)
+		<-unaryDone.Done()
+		var tasiSent traceAndSpanIDString
+		for _, tasi := range traceAndSpanIDs {
+			if strings.HasPrefix(tasi.spanName, "Sent.") {
+				tasiSent = idsToString(tasi, projectID)
+				continue
+			}
+		}
+
+		fle.mu.Lock()
+		for _, tasiSeen := range fle.idsSeen {
+			if diff := cmp.Diff(tasiSeen, &tasiSent, cmp.AllowUnexported(traceAndSpanIDString{})); diff != "" {
+				readerErrCh.Send(errors.New("got unexpected id, should be a client span"))
+			}
+		}
+
+		fle.entries = nil
+		fle.mu.Unlock()
+		readerErrCh.Send(nil)
+	}()
+	if _, err := ss.Client.UnaryCall(ctx, &grpc_testing.SimpleRequest{Payload: &grpc_testing.Payload{Body: testOkPayload}}); err != nil {
+		t.Fatalf("Unexpected error from UnaryCall: %v", err)
+	}
+	unaryDone.Fire()
+	if chErr, err := readerErrCh.Receive(ctx); chErr != nil || err != nil {
+		if err != nil {
+			t.Fatalf("Should have received something from error channel: %v", err)
+		}
+		if chErr != nil {
+			t.Fatalf("Should have received a nil error from channel, instead received: %v", chErr)
+		}
+	}
+	wg.Wait()
```

Comment:
Is this waitgroup important?  The goroutine ends with a push onto the readerErrCh, which is your signal that it's done anyway.  I think wg can just be deleted.

Reply:
> So the cleanup guarantee is the last operation of the goroutine happens before the main goroutine exits?

What do you mean by "cleanup guarantee"?

To satisfy the leak checker, all goroutines started by a test (directly or indirectly) need to terminate within 10 seconds (I believe, but the goal would be as close to 0 seconds a possible).